File: tests/c/matrix/test_benchmark.c
#include <stdio.h>
#include <stdlib.h>
#include "../utils/utils.h"

void matrix_multiplication(const double *A, const double *B, double *C, int n);

int main(int argc, char *argv[]) {
    const char *size = "small";
    if (argc >= 2) size = argv[1];

    char path[256];
    snprintf(path, sizeof(path), "../../../datasets/benchmark/matrix/%s.bin", size);

    MatrixData *md = load_matrix_data(path);
    printf("Matrix Multiplication\n");
    printf("Dataset: %s\n", path);
    printf("Size: %dx%d\n", md->n, md->n);

    double *C = (double *)calloc(md->n * md->n, sizeof(double));
    if (C == NULL) {
        fprintf(stderr, "malloc failed\n");
        return 1;
    }

    matrix_multiplication(md->A, md->B, C, md->n);

    printf("A (top-left corner):\n");
    print_matrix(md->A, md->n);

    printf("B (top-left corner):\n");
    print_matrix(md->B, md->n);

    printf("C = A*B (top-left corner):\n");
    print_matrix(C, md->n);

    /* C[0][0] should equal dot product of row 0 of A and col 0 of B */
    double expected = 0.0;
    for (int k = 0; k < md->n; k++) {
        expected += md->A[k] * md->B[k * md->n];
    }
    printf("C[0][0] check: %.4f (computed) vs %.4f (expected) \n%s\n", C[0], expected,
    (C[0] - expected < 1e-6 && C[0] - expected > -1e-6) ? "OK" : "FAIL");

    double checksum = matrix_checksum(C, md->n);

    printf("RESULT {\"n\":%d,\"c00\":%.4f,\"checksum\":%.4f}\n", md->n, C[0], checksum);
    free(C);
    free_matrix_data(md);
    return 0;
}

File: tests/js/matrix/test_benchmark.mjs
import { loadMatrixData } from '../test_loaders.mjs';
import { matrix_multiplication } from '../../../src/js/numeric/matrix_multiplication.mjs';
import { matrixChecksum } from '../../../src/js/utils/utils.mjs';

const size = process.argv[2] || 'small';
const path = `../../../datasets/benchmark/matrix/${size}.bin`;
const matrixData = loadMatrixData(inputPath);

const n = matrixData.n;

matrix_multiplication(matrixData.A, matrixData.B, matrixData.C, n);

let expected = 0.0;
for (let k = 0; k < n; k++) {
    expected += matrixData.A[k] * matrixData.B[k * n];
}

const diff = Math.abs(matrixData.C[0] - expected);

console.log('Matrix multiplication');
console.log(`Dataset: ${path}`);
console.log(`Size: ${n}x${n}`);
console.log(`C[0][0] check: ${matrixData.C[0].toFixed(4)} (computed) vs ${expected.toFixed(4)} (expected)`);
console.log(diff < 1e-6 ? 'OK' : 'FAIL');

console.log("RESULT", JSON.stringify({
    n,
    c00: Number(matrixData.C[0].toFixed(4)),
    checksum: Number(matrixChecksum(matrixData.C, n).toFixed(4))
}));